### PR TITLE
Handle authentication errors in network client

### PR DESCRIPTION
**Pull request by stel**
Currently only errors that could be returned from `SyncUser.authenticate...` are handled:
- `RLMSyncAuthErrorInvalidCredential`
- `RLMSyncAuthErrorUserDoesNotExist`
- `RLMSyncAuthErrorUserAlreadyExists`

Any other non-successful HTTP codes are still treated as `RLMSyncErrorHTTPStatusCodeError`, but may contain specific error messages from server. This restriction could be removed by defining additional error codes.

I think error reporting needs to be discussed, my idea is to wrap all the network, parsing, etc errors into sync-specific ones, it doesn't make much sense for users to get `NSCocoaErrorDomain` errors from sync api calls in most cases, but this errors still could be included by `NSUnderlyingErrorKey` into `RLMSyncErrorDomain` errors. Currently it's not a part of this PR.

/cc @austinzheng, @jpsim

### CODE REVIEW

**Comment by austinzheng** [comment]
Can you rebase this onto `master` or resolve the conflicts?

File: Realm/ObjectServerTests/RLMObjectServerTests.m [hunk 1]
```diff
@@ -45,32 +45,85 @@ - (void)testUsernamePasswordAuthentication {
...
+        XCTAssertEqual(error.domain, RLMSyncErrorDomain);
+        XCTAssertEqual(error.code, RLMSyncAuthErrorUserDoesNotExist);
+        XCTAssertNotNil(error.localizedDescription);
+
+        [expectation fulfill];
+    }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+}
+
+/// Registereing a user with existing username should return corresponding error.
```
Comment (austinzheng) [inline Realm/ObjectServerTests/RLMObjectServerTests.m]: "Registering"

File: Realm/RLMNetworkClient.m [hunk 2]
```diff
@@ -156,4 +150,80 @@ + (void)postRequestToEndpoint:(RLMServerEndpoint)endpoint
...
+            // Fallback to HTTP status error without any additional info
+            *error = [NSError errorWithDomain:RLMSyncErrorDomain
+                                         code:RLMSyncErrorHTTPStatusCodeError
+                                     userInfo:@{kRLMSyncErrorStatusCodeKey: @(httpResponse.statusCode)}];
+        }
+
+        return NO;
+    }
+
+    if (data == nil) {
```
Comment (austinzheng) [inline Realm/RLMNetworkClient.m]: I think our convention is just `if (!data) { ...`.

File: Realm/RLMNetworkClient.m [hunk 3]
```diff
@@ -156,4 +150,80 @@ + (void)postRequestToEndpoint:(RLMServerEndpoint)endpoint
...
+
+    id json = [NSJSONSerialization JSONObjectWithData:data
+                                              options:(NSJSONReadingOptions)0
+                                                error:nil];
+    if (!json || ![json isKindOfClass:[NSDictionary class]]) {
+        return nil;
+    }
+
+    RLMSyncErrorResponseModel *responseModel = [[RLMSyncErrorResponseModel alloc] initWithDictionary:json];
+    if (responseModel == nil) {
```
Comment (austinzheng) [inline Realm/RLMNetworkClient.m]: Same

**Comment by stel** [comment]
Thanks, @austinzheng! Master is merged now

**Comment by stel** [comment]
bump

**Comment by stel** [comment]
@austinzheng, @jpsim any other feedback for this?

**Comment by stel** [comment]
Awesome, thanks! Merging now.

**Pull request merged**